File: dev/R/R/parser.R
xml_text_first <- function(node, xpath) {
  values <- XML::xpathSApply(node, xpath, XML::xmlValue)
  if (length(values) == 0) {
    return("")
  }
  trimws(gsub("\r\n?", "\n", values[[1]]))
}

xml_attr_first <- function(node, xpath, attr) {
  found <- XML::getNodeSet(node, xpath)
  if (length(found) == 0) {
    return("")
  }
  value <- XML::xmlGetAttr(found[[1]], attr, default = "")
  if (is.na(value)) "" else value
}

calculated_property <- function(drug_node, kind) {
  property_nodes <- XML::getNodeSet(drug_node, "calculated-properties/property")

  for (property_node in property_nodes) {
    if (identical(xml_text_first(property_node, "kind"), kind)) {
      return(xml_text_first(property_node, "value"))
    }
  }
  ""
}

parse_drugbank_xml <- function(path, profile = "core", modules = NULL, schema_dir = NULL) {
  if (!file.exists(path)) {
    stop(sprintf("Input XML file does not exist: %s", path), call. = FALSE)
  }

  selected_modules <- resolve_modules(
    profile = profile,
    modules = modules,
    schema_dir = schema_dir
  )
  tables <- resolve_tables(selected_modules, schema_dir = schema_dir)
  result <- new_parse_result(tables)

  if (length(tables) == 0) {
    return(unname(result))
  }

  if ("core" %in% selected_modules) {
    handle_drug <- function(drug_node, ...) {
      result <<- extract_core_drug(drug_node, result)
      FALSE
    }
    XML::xmlEventParse(
      path,
      branches = list(drug = handle_drug),
      useTagName = TRUE,
      addContext = FALSE,
      trim = FALSE
    )
  }

  deduplicate_table(result, "targets", key_fields = "target_id")
}

extract_core_drug <- function(drug_node, result) {
  drug_id <- xml_text_first(drug_node, "drugbank-id[@primary='true']")
  if (identical(drug_id, "")) {
    return(result)
  }

  drug_name <- xml_text_first(drug_node, "name")
  indication <- xml_text_first(drug_node, "indication")
  inchi <- calculated_property(drug_node, "InChI")

  result <- append_row(result, "drugs", list(
    drug_id = drug_id,
    drug_name = drug_name,
    inchi = inchi,
    source = "DrugBank"
  ))
  result <- append_row(result, "drug_indication", list(
    drug_id = drug_id,
    indication = indication,
    source = "DrugBank"
  ))

  target_nodes <- XML::getNodeSet(drug_node, "targets/target")
  for (target_node in target_nodes) {
    target_id <- xml_attr_first(target_node, "polypeptide", "id")
    if (identical(target_id, "")) {
      next
    }

    target_name <- xml_text_first(target_node, "name")
    gene_name <- xml_text_first(target_node, "polypeptide/gene-name")
    organism <- xml_text_first(target_node, "organism")

    result <- append_row(result, "targets", list(
      target_id = target_id,
      target_name = target_name,
      gene_name = gene_name,
      organism = organism,
      source = "DrugBank"
    ))
    result <- append_row(result, "drug_target", list(
      drug_id = drug_id,
      target_id = target_id,
      source = "DrugBank"
    ))
    result <- append_row(result, "target_drug_indication", list(
      target_id = target_id,
      gene_name = gene_name,
      drug_id = drug_id,
      drug_name = drug_name,
      inchi = inchi,
      indication = indication,
      source = "DrugBank"
    ))
  }

  result
}

deduplicate_table <- function(result, table, key_fields) {
  if (!table %in% names(result) || nrow(result[[table]]) == 0) {
    return(result)
  }

  keys <- do.call(paste, c(result[[table]][key_fields], sep = "\r"))
  result[[table]] <- result[[table]][!duplicated(keys), , drop = FALSE]
  rownames(result[[table]]) <- NULL
  result
}

File: dev/python/drugbank_parse/parser.py
from __future__ import annotations

from pathlib import Path

from lxml import etree

from .models import ParseResult
from .profiles import resolve_modules, resolve_tables

DRUGBANK_NS = "http://www.drugbank.ca"
NS = {"db": DRUGBANK_NS}
SOURCE = "DrugBank"


def parse_drugbank_xml(
    path: str | Path,
    profile: str = "core",
    modules: list[str] | None = None,
) -> ParseResult:
    xml_path = Path(path)
    if not xml_path.exists():
        raise FileNotFoundError(f"Input XML file does not exist: {xml_path}")

    selected_modules = resolve_modules(profile=profile, modules=modules)
    tables = resolve_tables(selected_modules)
    result = ParseResult(tables={table: [] for table in tables})

    context = etree.iterparse(
        str(xml_path),
        events=("end",),
        tag=f"{{{DRUGBANK_NS}}}drug",
        recover=False,
    )

    for _, drug_node in context:
        if "core" in selected_modules:
            _extract_core_drug(drug_node, result)
        while drug_node.getprevious() is not None:
            del drug_node.getparent()[0]
        drug_node.clear()

    if "targets" in result.tables:
        _deduplicate_table(result, "targets", key_fields=("target_id",))
    return result


def _extract_core_drug(drug_node: etree._Element, result: ParseResult) -> None:
    drug_id = _first_text(drug_node, "db:drugbank-id[@primary='true']")
    if not drug_id:
        return

    drug_name = _first_text(drug_node, "db:name")
    indication = _first_text(drug_node, "db:indication")
    inchi = _calculated_property(drug_node, "InChI")

    drug_row = {
        "drug_id": drug_id,
        "drug_name": drug_name,
        "inchi": inchi,
        "source": SOURCE,
    }
    result.add_row("drugs", drug_row)
    result.add_row(
        "drug_indication",
        {
            "drug_id": drug_id,
            "indication": indication,
            "source": SOURCE,
        },
    )

    for target_node in drug_node.xpath("db:targets/db:target", namespaces=NS):
        target_id = _target_id(target_node)
        if not target_id:
            continue

        target_name = _first_text(target_node, "db:name")
        organism = _first_text(target_node, "db:organism")
        gene_name = _first_text(target_node, "db:polypeptide/db:gene-name")

        result.add_row(
            "targets",
            {
                "target_id": target_id,
                "target_name": target_name,
                "gene_name": gene_name,
                "organism": organism,
                "source": SOURCE,
            },
        )
        result.add_row(
            "drug_target",
            {
                "drug_id": drug_id,
                "target_id": target_id,
                "source": SOURCE,
            },
        )
        result.add_row(
            "target_drug_indication",
            {
                "target_id": target_id,
                "gene_name": gene_name,
                "drug_id": drug_id,
                "drug_name": drug_name,
                "inchi": inchi,
                "indication": indication,
                "source": SOURCE,
            },
        )


def _target_id(target_node: etree._Element) -> str:
    polypeptide = target_node.find("db:polypeptide", namespaces=NS)
    if polypeptide is not None:
        return polypeptide.get("id", "") or ""
    return ""


def _first_text(node: etree._Element, xpath: str) -> str:
    values = node.xpath(xpath, namespaces=NS)
    if not values:
        return ""
    value = values[0]
    if isinstance(value, etree._Element):
        return " ".join(value.itertext()).strip()
    return str(value).strip()


def _calculated_property(drug_node: etree._Element, kind: str) -> str:
    properties = drug_node.xpath("db:calculated-properties/db:property", namespaces=NS)
    for property_node in properties:
        property_kind = _first_text(property_node, "db:kind")
        if property_kind == kind:
            return _first_text(property_node, "db:value")
    return ""


def _deduplicate_table(
    result: ParseResult,
    table: str,
    key_fields: tuple[str, ...],
) -> None:
    seen = set()
    rows = []
    for row in result.rows(table):
        key = tuple(row.get(field, "") for field in key_fields)
        if key in seen:
            continue
        seen.add(key)
        rows.append(row)
    result.tables[table] = rows
